internal/luks2/metadata.go

```diff
@@ -0,0 +1,910 @@
+// of modifications by libcryptsetup will be detected by the checksum verification.
+//
+// Note that this function does not attempt recovery of either header in the event that one of the
+// headers is not valid - we leave this to libcryptsetup, which happens automatically on any
+// cryptsetup or systemd-cryptsetup invocation. This package does not directly perform any
+// modifications to the header.
+//
+// This function requires an advisory shared lock on the LUKS container associated with the
+// specified path. If the mode parameter is LockModeBlocking, this function will block until the
+// lock can be obtained. If the mode parameter is LockModeNonBlocking, a wrapped syscall.Errno
+// error with the value of syscall.EWOULDBLOCK will be returned if the lock can not be obtained.
+func ReadHeader(path string, lockMode LockMode) (*HeaderInfo, error) {
+	releaseLock, err := acquireSharedLock(path, lockMode)
+	if err != nil {
+		return nil, xerrors.Errorf("cannot acquire shared lock: %w", err)
+	}
+	defer releaseLock()
+
+	f, err := os.Open(path)
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+
+	// Try to decode and check the primary header
+	primaryHdr, primaryJSONData, primaryErr := decodeAndCheckHeader(f, 0, true)
+	var primaryMetadata Metadata
+	if primaryErr == nil {
+		if err := json.NewDecoder(primaryJSONData).Decode(&primaryMetadata); err != nil {
+			primaryErr = xerrors.Errorf("cannot decode JSON metadata area: %w", err)
+		}
+	}
+
+	var secondaryHdr *binaryHdr
+	var secondaryJSONData *bytes.Buffer
+	var secondaryErr error
+	if primaryErr != nil {
+		// No valid primary header. Try to decode and check a secondary header from one of the
+		// well known offsets (see Table 1: Possible LUKS2 secondary header offsets and JSON area
+		// size in the LUKS2 On-Disk Format specification).
+		for _, off := range []int64{0x4000, 0x8000, 0x10000, 0x20000, 0x40000, 0x80000, 0x100000, 0x200000, 0x400000} {
+			secondaryHdr, secondaryJSONData, secondaryErr = decodeAndCheckHeader(f, off, false)
+			if secondaryErr == nil {
+				break
+			}
+		}
+	} else {
+		// Try to decode and check the secondary header immediately after the primary header.
+		secondaryHdr, secondaryJSONData, secondaryErr = decodeAndCheckHeader(f, int64(primaryHdr.HdrSize), false)
+	}
+	var secondaryMetadata Metadata
+	if secondaryErr == nil {
+		if err := json.NewDecoder(secondaryJSONData).Decode(&secondaryMetadata); err != nil {
+			secondaryErr = xerrors.Errorf("cannot decode JSON metadata area: %w", err)
+		}
+	}
+
+	var hdr *binaryHdr
+	var metadata *Metadata
+	switch {
+	case primaryErr == nil && secondaryErr == nil:
+		// Both headers are valid
+		hdr = primaryHdr
+		metadata = &primaryMetadata
+		switch {
+		case secondaryHdr.SeqId < primaryHdr.SeqId:
+			// The secondary header is obsolete. Cryptsetup will recover this automatically.
+			fmt.Fprintf(os.Stderr, "luks2.ReadHeader: secondary header for %s is obsolete\n", path)
```

Comment:
here and elsewhere you could use something like:

var osStderr = os.Stderr

then in the tests replace it with a buffer,  this avoids spurious output from the tests and also can allow to test what gets printed if desired. At the moment running the tests prints at least this:

```
luks2.ReadHeader: primary header for /tmp/check-5656277456882226488/16/luks2-hdr2-invalid-checksum0.img is invalid: invalid header checksum
luks2.ReadHeader: primary header for /tmp/check-5656277456882226488/20/luks2-hdr-invalid-checksum0.img is invalid: invalid header checksum
luks2.ReadHeader: secondary header for /tmp/check-5656277456882226488/22/luks2-hdr-invalid-checksum1.img is invalid: invalid header checksum
luks2.ReadHeader: primary header for /tmp/check-5656277456882226488/26/luks2-hdr-obsolete0.img is obsolete
```